### PR TITLE
Add a Makefile rule to run cargo-bloat

### PR DESCRIPTION
**Pull request by hudson-ayers**
### Pull Request Overview

Now, cargo-bloat can be run by running `make cargobloat` from any
`boards/xx` directory. This is useful for identifying large functions
in the kernel binary. To pass additional flags to the invocation of
cargo-bloat, set the BLOAT_FLAGS environment variable. It also adds
a Makefile rule to run cargo-bloat on a binary built with inlining
disabled (as much as possible) which can allow for more accurate
attribution of size.

The new `CARGO_FLAGS_TOCK_NO_BUILD_STD` variable ensures that
cargo-bloat works even when this rule is called from an out-of-tree
board that passes the `-Z build-std` flag to `CARGO_FLAGS_TOCK`. This is
necessary because cargo-bloat does not support that flag.

This change automatically installs cargo-bloat when this rule is run. As a result the first invocation might take awhile.

Example outputs for Imix:
```
hudsonayers: ~/upstream_tock/tock/boards/imix (cargo-bloat) $ make cargobloat
    Finished release [optimized + debuginfo] target(s) in 0.04s
    Analyzing target/thumbv7em-none-eabi/release/imix

File  .text     Size      Crate Name
0.2%   4.7%   8.2KiB  [Unknown] main
0.2%   4.1%   7.3KiB     kernel kernel::sched::Kernel::kernel_loop
0.1%   1.5%   2.6KiB   capsules capsules::net::ipv6::ipv6_send::IP6SendStruct<A>::send_next_fragment
0.1%   1.4%   2.6KiB  capsules? <capsules::adc::AdcDedicated<A> as kernel::driver::Driver>::command
0.1%   1.4%   2.5KiB   capsules capsules::net::sixlowpan::sixlowpan_compression::decompress
0.0%   1.2%   2.2KiB components capsules::process_console::ProcessConsole<C>::write_state
0.0%   1.2%   2.1KiB components <capsules::process_console::ProcessConsole<C> as kernel::hil::uart::TransmitClient>::transmitted_buffer
0.0%   1.2%   2.0KiB     kernel kernel::process_utilities::load_processes
0.0%   1.1%   2.0KiB   capsules capsules::net::ieee802154::Header::decode
0.0%   1.1%   1.9KiB   capsules <capsules::ieee802154::framer::Framer<M,A> as capsules::ieee802154::device::MacDevice>::prepare_data_frame
0.0%   1.1%   1.9KiB   capsules <capsules::net::udp::driver::UDPDriver as kernel::driver::Driver>::command
0.0%   1.0%   1.8KiB  capsules? <capsules::rf233::RF233<S> as kernel::hil::spi::SpiMasterClient>::read_write_done
0.0%   1.0%   1.7KiB   capsules <capsules::net::sixlowpan::sixlowpan_state::Sixlowpan<A,C> as capsules::ieee802154::device::RxClient>::receive
0.0%   0.8%   1.4KiB     kernel kernel::debug::panic_print
0.0%   0.7%   1.3KiB  capsules? <capsules::virtual_aes_ccm::MuxAES128CCM<A> as kernel::hil::symmetric_encryption::Client>::crypt_done
0.0%   0.7%   1.2KiB  capsules? <capsules::usb::usbc_client::Client<C> as kernel::hil::usb::Client>::ctrl_setup
0.0%   0.7%   1.2KiB        std <char as core::fmt::Debug>::fmt
0.0%   0.7%   1.2KiB  capsules? <capsules::adc::AdcDedicated<A> as kernel::hil::adc::HighSpeedClient>::samples_ready
0.0%   0.6%   1.0KiB     kernel <kernel::process_standard::ProcessStandard<C> as kernel::process::Process>::print_full_process
0.0%   0.6%   1.0KiB   capsules <capsules::net::ipv6::ipv6_recv::IP6RecvStruct as capsules::net::sixlowpan::sixlowpan_state::SixlowpanRxClient>::receive
2.1%  51.7%  91.0KiB            And 1058 smaller methods. Use -n N to show more.
4.0% 100.0% 176.2KiB            .text section size, the file size is 4.3MiB
```

```
hudsonayers: ~/upstream_tock/tock/boards/imix (cargo-bloat) $ BLOAT_FLAGS='--crates' make cargobloat
    Finished release [optimized + debuginfo] target(s) in 0.04s
    Analyzing target/thumbv7em-none-eabi/release/imix

File  .text     Size Crate
1.5%  38.6%  68.0KiB capsules
0.5%  13.2%  23.2KiB kernel
0.3%   7.8%  13.8KiB std
0.3%   7.2%  12.6KiB [Unknown]
0.2%   5.2%   9.1KiB sam4l
0.2%   4.4%   7.7KiB components
0.1%   1.5%   2.7KiB cortexm
0.0%   0.5%     876B imix
0.0%   0.1%     262B tock_cells
0.0%   0.1%     104B tock_tbf
4.0% 100.0% 176.2KiB .text section size, the file size is 4.3MiB

Note: numbers above are a result of guesswork. They are not 100% correct and never will be.
```

### Testing Strategy

This pull request was tested by running `make cargobloat` in the Imix/ directory.

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] No updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by alistair23 — APPROVED** [review]

File: boards/Makefile.common [hunk 1]
```diff
@@ -281,6 +281,38 @@ lst: $(TARGET_PATH)/release/$(PLATFORM).lst
 show-target:
 	$(info $(TARGET))
 
+# Additional flags that can be passed to cargo bloat via an
+# environment variable. Allows users to pass arbitrary flags
+# supported by cargo bloat to customize the output. By default,
+# pass an empty string.
+BLOAT_FLAGS ?=
```
Comment (ppannuto) [inline boards/Makefile.common:288]:
```suggestion
CARGO_BLOAT_FLAGS ?=
```
?

**Review by ppannuto — COMMENTED** [review]
Should we add some documentation on this somewhere for discoverability? Maybe as one of the options that prints when the empty root makefile runs?

File: boards/Makefile.common [hunk 2]
```diff
@@ -281,6 +281,38 @@ lst: $(TARGET_PATH)/release/$(PLATFORM).lst
...
+
+
+# cargo bloat does not support -Z build-std, so we must remove it from cargo flags
+# see https://github.com/RazrFalcon/cargo-bloat/issues/62
+CARGO_FLAGS_TOCK_NO_BUILD_STD := $(filter-out -Z build-std=core,$(CARGO_FLAGS_TOCK))
```
Comment (bradjc) [inline boards/Makefile.common:293]: These should be moved up in the file, and should be included in our debug output.

**Comment by hudson-ayers** [comment]
> Should we add some documentation on this somewhere for discoverability? Maybe as one of the options that prints when the empty root makefile runs?

I added documentation for both this rule and the stack-analysis rule as one of the options that prints when the empty root makefile runs

File: Makefile [hunk 3]
```diff
@@ -36,6 +36,12 @@ usage:
 	@echo "and usually 'make program' or 'make flash' to load Tock onto hardware."
 	@echo "Check out the README in your board's folder for more information."
 	@echo
+	@echo "There are a few helpful targets that can be run for all individual boards."
+	@echo "To run these, run 'make {target} from the board directory for any of the"
```
Comment (bradjc) [inline Makefile:40]:
```suggestion
	@echo "To run these, run 'make {target}' from the board directory for any of the"
```

**Review by ppannuto — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]
Looks good, although in line with the general principle of "don't install anything without asking" (given that `cargo install` will install the cargo-bloat bin user-wide), should we maybe prompt the user first?

**Comment by hudson-ayers** [comment]
> Looks good, although in line with the general principle of "don't install anything without asking" (given that `cargo install` will install the cargo-bloat bin user-wide), should we maybe prompt the user first?

We still automatically install llvm-tools if they are not installed (e.g. llvm-objdump, llvm-size, etc.) and automatically add rustup  targets without a prompt. If you want to add a commit to prompt a user I am happy to accept it, but I don't really feel it is necessary, especially since the download only occurs if you try to run a `make` rule specifically targeted at running this tool. It's not as though `cargo-bloat` will be installed as part of compiling a board, for example. And `cargo-bloat` is not run by any of the CI checks either.

**Review by alistair23 — APPROVED** [review]

**Comment by lschuermann** [comment]
@hudson-ayers Fair. Not a fan, but I agree this is in line with how we do things currently. :)

**Comment by bradjc** [comment]
I think running `make cargobloat` is pretty clear on the intention, and we should just do what the user wants. If the user wants more control they can run cargo bloat directly.

**Pull request merged**